File: code/com/leetcode/solutions/NO_0011_盛最多水的容器/Solution.java
package com.leetcode.solutions.NO_0011_盛最多水的容器;

/**
 * Created by gang.xu01@hand-china.com on 2019/1/6
 */
public class Solution {

    /**
     * 解法：指针满足比长度小的条件便从两边向中间移动，直到取到最大值
     *
     * @param height
     * @return
     */
    public int maxArea(int[] height) {

        int left = 0,
            right = height.length - 1,
            area = 0,
            l, r;

        while (left < right) {
            area = Math.max((right - left) * Math.min(height[left], height[right]), area);
            l = height[left];
            r = height[right];
            if (l < r) while (l >= height[left] && left < right) left++;
            else while (r >= height[right] && left < right) right--;
        }

        return area;
    }

    public static void main(String[] args) {
        new Solution().maxArea(new int[]{1,8,6,2,5,4,8,3,7});
    }
}

File: code/com/leetcode/solutions/NO_0011_盛最多水的容器/Solution.js
/**
 * 思路：设置左右指针向中靠拢，因最大容器取决于最短边，故较短边优先向另一方向移动
 * @param {number[]} height
 * @return {number}
 */
var maxArea = function (height) {
    let left = 0,
        right = height.length - 1,
        maxSize = 0,
        a, b;

    while (left < right) {
        maxSize = Math.max((right - left) * Math.min(height[left], height[right]), maxSize);
        a = height[left], b = height[right];
        if (a > b) {
            while (height[right] <= b) {
                right--;
            }
        } else {
            while (height[left] <= a) {
                left++;
            }
        }
    }
    return maxSize;
};